### DISCUSSION TITLE
Using hipSYCL with OneAPI libraries

### DISCUSSION THREAD

**Discussion by mspronesti** [AdaptiveCpp/AdaptiveCpp#753] [Q&A]
Hello, 
is it possible to compile a project which uses Intel libraries like oneTBB and oneMKL with hipSYCL ? 
I tried compiling the entire toolchain with syclcc to no avail. 
Actually, they don't even seem compatible to me in terms of SYCL version.
Thanks in advance!

**Comment by illuhad** [comment]
Hi!

I don't think anybody has looked into oneTBB yet. I don't know what exactly it does internally, so I can't say if it might work.
However, upstream oneMKL at https://github.com/oneapi-src/oneMKL does have hipSYCL support for its BLAS domains on CPU, NVIDIA GPUs and AMD GPUs and RNG domains at least on AMD GPUs.
You will need to build oneMKL using hipSYCL, and set some cmake variables when building oneMKL. @sbalint98, @nilsfriess or @normallytangent can provide details.

What do you mean by being incompatible in terms of SYCL version?

**Reply by mspronesti** [reply]
Hi!
Thanks a lot for your answer. As far as I know oneMKL depends on oneTBB so I don't get how oneMKL Is supported.
Anyway I tried compiling It myself and I didn't menage to.
Going to the second part, I tried compiling some sycl code (which doesn't use any Intel's library) and I get a bunch of errors related to different declarations (e.g. `sycl::plus<>` type not being defaulted, everything in `cl::sycl` instead of `sycl::` , some methods in `cl::sycl::nd_range` not found ).
Sorry if I miss something obvious and thanks again for your help.

**Reply by nilsfriess** [reply]
Could you provide more details about your setup and how you are trying to build oneMKL with hipSYCL? Did you follow the steps here: https://oneapi-src.github.io/oneMKL/building_the_project.html#building-for-rocm-with-hipsycl ? If yes, what exactly is not working?

**Reply by illuhad** [reply]
> Thanks a lot for your answer. As far as I know oneMKL depends on oneTBB so I don't get how oneMKL Is supported.

oneMKL supports oneTBB as a potential threading backend for CPU. We are relying on oneMKL's SYCL support, not TBB.

> Going to the second part, I tried compiling some sycl code (which doesn't use any Intel's library) and I get a bunch of errors related to different declarations (e.g. sycl::plus<> type not being defaulted, everything in cl::sycl instead of sycl:: , some methods in cl::sycl::nd_range not found ).

I can give you more specific answers if you can show the code that causes the issue, but are you perhaps including the deprecated `<CL/sycl.hpp>` header instead of the proper SYCL 2020 `<sycl/sycl.hpp>` header? There is an ambiguity in the SYCL 2020 specification regarding the legacy `CL/sycl.hpp` header. It says, that when including `CL/sycl.hpp`, the SYCL interface should be exposed in `::cl::sycl`. It does however not say whether it should *additionally* also be exposed in `::sycl` when using this header. DPC++ however does that, while hipSYCL does not follow that interpretation.
The other errors might be due to other implementation differences or DPC++ extensions that you might be using, but it's difficult to guess without seeing the code.

**Reply by mspronesti** [reply]
Hi, 
sorry for my late reply: I've been busy with my exams.
@illuhad thanks a lot for your clarification, especially the part related to the difference in the headers between DPC++ and hipSYCL. It solved the majority of my problems.
@nilsfriess thanks a lot, I was already following that guide, but using the wrong header caused plenty of troubles when including `oneapi/mkl.hpp`. 
Now everything is solved, except the fact that for some "standard sycl functions" I notice some differences. For instance, consider this snippet

```C++
ulong aggregate = sycl::reduce_over_group(item.get_group(), sum, sycl::plus<>());
```
which doesn't compile if I don't specify `ulong` as template type for `sycl::plus`. It however works when I compile this code using `dpcpp` or `clang++` (llvm) .

**Reply by illuhad** [reply]
This is due to a change in the specification from SYCL 2020 provisional to SYCL 2020. I believe it is addressed in PR #778.
